File: src/lib/github.ts
import { graphql } from '@octokit/graphql';
import { GraphQLResponse } from '@/types/github';

const GITHUB_GRAPHQL_QUERY = `
query GetDevWrappedData($username: String!) {
  user(login: $username) {
    name
    login
    avatarUrl
    bio
    followers {
      totalCount
    }
    following {
      totalCount
    }
    contributionsCollection {
      totalIssueContributions
      totalCommitContributions
      totalPullRequestContributions
      totalPullRequestReviewContributions
      contributionCalendar {
        totalContributions
        weeks {
          contributionDays {
            date
            contributionCount
          }
        }
      }
    }
    pullRequests(first: 100, states: [MERGED, CLOSED, OPEN], orderBy: { field: CREATED_AT, direction: DESC }) {
      totalCount
      nodes {
        state
      }
    }
    repositories(first: 50, ownerAffiliations: OWNER, isFork: false, orderBy: { field: UPDATED_AT, direction: DESC }) {
      nodes {
        name
        stargazerCount
        isFork
        languages(first: 5, orderBy: {field: SIZE, direction: DESC}) {
          edges {
            size
            node {
              name
              color
            }
          }
        }
      }
    }
  }
}
`

export async function fetchRawGithubData(username: string): Promise<GraphQLResponse> {
    const token = process.env.GITHUB_TOKEN;
    if (!token) {
        throw new Error('GITHUB_TOKEN is not set in the environment variables.');
    }

    const graphqlClient = graphql.defaults({
        headers: {
            authorization: `token ${token}`
        }
    });

    const client = graphql.defaults({
      headers: {
        authorization: `bearer ${token}`,
      },
    });

    return await client<GraphQLResponse>(GITHUB_GRAPHQL_QUERY, { username });
}

File: go-engine/domain/github.go
package domain

import "github.com/shurcooL/githubv4"

// GraphQLQuery defines the exact payload we need from GitHub in a single round-trip.
type GraphQLQuery struct {
	User struct {
		Name      githubv4.String
		Login     githubv4.String
		AvatarURL githubv4.URI
		Bio       githubv4.String
		Followers struct {
			TotalCount githubv4.Int
		}
		Following struct {
			TotalCount githubv4.Int
		}
		ContributionsCollection struct {
			TotalIssueContributions             githubv4.Int
			TotalCommitContributions            githubv4.Int
			TotalPullRequestContributions       githubv4.Int
			TotalPullRequestReviewContributions githubv4.Int
			ContributionCalendar                struct {
				TotalContributions githubv4.Int
				Weeks              []struct {
					ContributionDays []struct {
						Date              githubv4.String
						ContributionCount githubv4.Int
					}
				}
			}
		}
		Repositories struct {
			Nodes []struct {
				Name           githubv4.String
				StargazerCount githubv4.Int
				IsFork         githubv4.Boolean
				Languages      struct {
					Edges []struct {
						Size githubv4.Int
						Node struct {
							Name  githubv4.String
							Color githubv4.String
						}
					}
				} `graphql:"languages(first: 10, orderBy: {field: SIZE, direction: DESC})"`
			}
		} `graphql:"repositories(first: 50, ownerAffiliations: OWNER, isFork: false, orderBy: {field: STARGAZERS, direction: DESC})"`
	} `graphql:"user(login: $username)"`
}
